### PR TITLE
Add support for zooming sub-plots

### LINKED ISSUE #13339: [FEATURE] Allow zoom tools to scale subplots/subcoordinates
### Problem description

The zoom tools operate only on the final target range on plots with subplots.

### Feature description

Allow the zoom tools to scale subcoordinate ranges on plots that contain subplots.

### Potential alternatives

None

### Additional information

<details>
<summary> Code </summary>

```
import numpy as np
from bokeh.plotting import figure
from bokeh.models import HoverTool, DataRange1d, FactorRange, Range1d, ColumnDataSource, WheelZoomTool, ZoomInTool, ZoomOutTool
from bokeh.io import show, output_notebook
from bokeh.core.properties import field
output_notebook()
np.random.seed(0)

n_channels = 10
n_seconds = 15
fs = 512
max_ch_disp = 5  # max channels to initially display
max_t_disp = 3 # max time in seconds to initially display

total_samples = n_seconds * fs
time = np.linspace(0, n_seconds, total_samples)
data = np.random.randn(n_channels, total_samples).cumsum(axis=1)
channels = [f'EEG {i}' for i in range(n_channels)]

hover = HoverTool(tooltips=[
    ("Channel", "$name"),
    ("Time", "$x s"),
    ("Amplitude", "$y µV"),
])

x_range = Range1d(start=time.min(), end=time.max())
#x_range = DataRange1d()
y_range = FactorRange(factors=channels)

p = figure(x_range=x_range, y_range=y_range, active_scroll="wheel_zoom", lod_threshold=None)
p.add_tools(hover)

source = ColumnDataSource(data=dict(time=time))
renderers = []

for i, channel in enumerate(channels):
    xy = p.subplot(
        x_source=p.x_range,
        y_source=Range1d(start=data[i].min(), end=data[i].max()),
        #y_source=DataRange1d(),
        x_target=p.x_range,
        y_target=Range1d(start=i, end=i + 1),
    )

    source.data[channel] = data[i]
    line = xy.line(field("time"), field(channel), color="black", source=source, name=channel)
    renderers.append(line)

p.add_tools(ZoomInTool(renderers=renderers))
p.add_tools(ZoomOutTool(renderers=renderers))

# p.add_tools(WheelZoomTool(renderers=renderers))

show(p)
```
</details>

<details>
<summary> Video of current behavior </summary>

https://github.com/bokeh/bokeh/assets/6613202/44eb977b-6229-4ee1-822c-b2714eec0f5a

</details>

<details>
<summary> Video of desired behavior (using different software) </summary>

https://github.com/bokeh/bokeh/assets/6613202/678f310f-39fb-48e2-956f-f1339f159c6f

</details>

### PR DESCRIPTION
**Pull request by mattpap**
This PR does two things:

1. It liberates range management internals from using named range/scale mappings, to make it possible for range management logic to work with sub-coordinates at all.
2. Makes various zoom tools and zooming logic aware of sub-coordinates and to scale accordingly.

Example. Notice that zoom in/out works on x frame range, however not y frame range, but y sub-coordinate ranges:

https://github.com/bokeh/bokeh/assets/27475/56cd3fdd-b99a-4994-acc4-dbf0952932e5

<details>
<summary>Code</summary>

```py
import numpy as np

from bokeh.core.properties import field
from bokeh.io import show
from bokeh.models import (ColumnDataSource, DataRange1d, FactorRange,
                          HoverTool, Range1d, WheelZoomTool, ZoomInTool, ZoomOutTool)
from bokeh.plotting import figure
from bokeh.palettes import Category10

np.random.seed(0)

n_channels = 10
n_seconds = 15
fs = 512
max_ch_disp = 5  # max channels to initially display
max_t_disp = 3 # max time in seconds to initially display

total_samples = n_seconds * fs
time = np.linspace(0, n_seconds, total_samples)
data = np.random.randn(n_channels, total_samples).cumsum(axis=1)
channels = [f'EEG {i}' for i in range(n_channels)]

hover = HoverTool(tooltips=[
    ("Channel", "$name"),
    ("Time", "$x s"),
    ("Amplitude", "$y µV"),
])

x_range = Range1d(start=time.min(), end=time.max())
#x_range = DataRange1d()
y_range = FactorRange(factors=channels)

p = figure(x_range=x_range, y_range=y_range, lod_threshold=None)

source = ColumnDataSource(data=dict(time=time))
renderers = []

for i, channel in enumerate(channels):
    xy = p.subplot(
        x_source=p.x_range,
        y_source=Range1d(start=data[i].min(), end=data[i].max()),
        #y_source=DataRange1d(),
        x_target=p.x_range,
        y_target=Range1d(start=i, end=i + 1),
    )

    source.data[channel] = data[i]
    line = xy.line(field("time"), field(channel), color=Category10[10][i], source=source, name=channel)
    renderers.append(line)

wheel_zoom = WheelZoomTool()#renderers=renderers)
zoom_in = ZoomInTool(renderers=renderers)
zoom_out = ZoomOutTool(renderers=renderers)

p.add_tools(wheel_zoom, zoom_in, zoom_out, hover)
p.toolbar.active_scroll = wheel_zoom

show(p)
```

</details>

TODO:

- [x] tests

fixes #13339

### CODE REVIEW

**Comment by bryevdv** [comment]
Really cool! Are `x_range` and `x_target` already included in any release, if so are they designated experimental? I think "source" is too generic and already overloaded, so that we should strongly consider alternative names for these properties to make searching for documentation more unambigious. E.g. maybe
*  `x_local`/ `x_remote` 
* `x_range` / `x_embed_range`

**Comment by droumis** [comment]
Looking great!

**Comment by mattpap** [comment]
This is tentatively ready for review. With this PR `ZoomInTool`, `ZoomOutTool` and `WheelZoomTool` were made aware of sub-coordinates. The default behavior is as before though more ranges will be taken into consideration. When working with true sub-coordinates, i.e. with `CompositeScale` in either or both dimensions, then the aforementioned tools provide new `level` property, which allows to configure which actual ranges will be zoomed. `0` is the default and acts on top-level (frame) ranges. Using positive values allows to descend composite scales and select inner sub-coordinate ranges.

**Comment by mattpap** [comment]
> Are x_range and x_target already included in any release, if so are they designated experimental? I think "source" is too generic and already overloaded, so that we should strongly consider alternative names for these properties to make searching for documentation more unambigious.

This should be discussed/performed in PR #13346. I'm generally not opposed to renaming. The current naming convention is taken from scales where we have `source_range` and `target_range`, thus `{x,y}_source` and `{x,y}_target` for composite scales.

**Comment by droumis** [comment]
Hi @mattpap , could you provide code for a working example? 

I'm getting `Javascript Error: unknown property WheelZoomTool.renderers` when trying to plot

**Comment by mattpap** [comment]
>  could you provide code for a working example?

@droumis, see the example (examples/interaction/tools/subcoordinates_zoom.py) added in this PR.

> I'm getting Javascript Error: unknown property WheelZoomTool.renderers when trying to plot

This is a mismatch between bokeh and bokehjs.

**Review by ianthomas23 — APPROVED** [review]
Looks good to me. It would be good to get @droumis' approval that this provides all the required functionality.

File: examples/interaction/tools/subcoordinates_zoom.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+on_change = CustomJS(
+    args=dict(tools=[wheel_zoom, zoom_in, zoom_out]),
+    code="""
+export default ({tools}, obj) => {
+    const level = obj.active ? 1 : 0
+    for (const tool of tools) {
+        tool.level = level
+    }
+}
+""")
```
Comment (droumis) [inline examples/interaction/tools/subcoordinates_zoom.py]:
```suggestion
ywheel_zoom = WheelZoomTool(renderers=renderers, level=level, dimensions="height")
xwheel_zoom = WheelZoomTool(renderers=renderers, level=level, dimensions="width")
zoom_in = ZoomInTool(renderers=renderers, level=level, dimensions="height")
zoom_out = ZoomOutTool(renderers=renderers, level=level, dimensions="height")

p.add_tools(ywheel_zoom, xwheel_zoom, zoom_in, zoom_out, hover)
p.toolbar.active_scroll = ywheel_zoom

on_change = CustomJS(
    args=dict(tools=[ywheel_zoom, zoom_in, zoom_out]),
    code="""
export default ({tools}, obj) => {
    const level = obj.active ? 1 : 0
    for (const tool of tools) {
        tool.level = level
    }
}
""")
```

The subplot zoom behavior for this particular example (and any others that I can think of) should only ever be dimension-constrained; it is undesirable to zoom in and out in the x-dim while simultaneously adjusting the y-dim scale of the individual subplots.

However, constraining the wheel zoom dimension to act only the y-dim prevents users from being able to hover over the x-axis to use x-dim constrained wheel zoom. Therefore, I've added an extra `xwheel_zoom` to retain this functionality, but have kept it inactive by default. Also, the behavior will be the same whether it's level 0 or 1 for this example so it is not included in the switch callback.

**Comment by droumis** [comment]
apart from my comment above, all looks good to go

File: src/bokeh/models/scales.py [hunk 2]
```diff
@@ -109,6 +111,26 @@ class CategoricalScale(Scale):
...
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+    source_scale = Required(Instance(Scale), help="""
+    The source scale.
+    """)
+
+    target_scale = Required(Instance(Scale), help="""
+    The target scale.
+    """)
```
Comment (bryevdv) [inline src/bokeh/models/scales.py:130]: These help strings don't add anything, really. What would be helpful for both of these is a brief concrete example to anchor the concepts, e.g. if we have a plot and a subplot, which one of those is the source and which one is the target?

Comment (mattpap) [reply src/bokeh/models/scales.py:130]: I will continue documenting sub-coordinates' related stuff in PR #13346.

**Review by droumis — APPROVED** [review]

**Pull request merged**